KIPs/kip-81.md

```diff
@@ -0,0 +1,773 @@
+        uint256 nodeBalance,
+        uint256 nodeVotes);
+
+    /// @dev Find the node who owns the given voter account
+    function getNodeFromVoter(address voter) external view returns(address nodeId);
+
+    /// @dev Find the voter account of the given node
+    function getVoterFromNode(address nodeId) external view returns(address voter);
+}
+```
+
+#### Usage
+
+A voting contract can utilize StakingTracker as follows.
+
+1. When a governance proposal is submitted, the voting contract calls `createTracker()` to finalize eligible GC nodes list and evaluate voting powers.
+2. Before `trackEnd` block, GCs stake or unstake their KLAYs from their CnStakingV2 contracts. The CnStakingV2 contracts will then call `refreshStake()` to notify balance change.
+3. GCs may change their voter account in their CnStakingV2 contracts. The CnStakingV2 contracts will then call `refreshVoter()` to notify voter account change.
+4. After `trackEnd` block, the voting powers are frozen. The voting contract use StakingTracker getters to process votes casted by voter accounts.
+
+#### Example implementation
+
+Below is an example implementation of `createTracker()`, `refreshStake()`, and `refreshVoter()` functions.
+
+```solidity
+abstract contract StakingTracker {
+    function createTracker(uint256 trackStart, uint256 trackEnd)
+        external override returns(uint256 trackerId)
+    {
+        trackerId = allTrackerIds.length + 1;
+        allTrackerIds.push(trackerId);
+
+        Tracker storage tracker = trackers[trackerId];
+        tracker.trackStart = trackStart;
+        tracker.trackEnd = trackEnd;
+        populateFromAddressBook(tracker);
+        calcAllVotes(tracker);
+        return trackerId;
+    }
+
+    function populateFromAddressBook(Tracker storage tracker) private {
+        (address[] memory nodeIds,
+         address[] memory stakingContracts,
+         address[] memory rewardAddrs, , ) = IAddressBook(ADDRESS_BOOK_ADDRESS()).getAllAddressInfo();
+
+        // Group staking contracts by common reward address
+        for (uint256 i = 0; i < nodeIds.length; i++) {
+            address n = nodeIds[i];
+            address s = stakingContracts[i];
+            address r = rewardAddrs[i];
+            if (tracker.rewardToNodeId[r] == address(0)) { // fresh rewardAddr
+                tracker.nodeIds.push(n);
+                tracker.rewardToNodeId[r] = n;
+            } else { // previously appeared rewardAddr
+                n = tracker.rewardToNodeId[r];
+            }
+            tracker.stakingToNodeId[s] = n;
+            tracker.stakingBalances[s] = s.balance;
+            tracker.nodeBalances[n] += s.balance;
+        }
+    }
+
+    function calcAllVotes(Tracker storage tracker) private {
+        tracker.totalVotes = 0;
+        tracker.eligibleNodes = 0;
+        for (uint256 i = 0; i < tracker.nodeIds.length; i++) {
+            if (isNodeEligible(trackerId, tracker.nodeIds[i])) {
+                tracker.eligibleNodes ++;
```

Comment:
we don't need to store every counter to the storage. Why don't you use a memory variable and store it to the storage after all the computation is done?

Reply:
Fixed in https://github.com/klaytn/kips/commit/acfd52ebad8d79fd4e300bc034cf23dc501cba96.